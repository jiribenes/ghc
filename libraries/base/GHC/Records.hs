{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Records
-- Copyright   :  (c) Adam Gundry 2015-2016
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- This module defines the 'HasField' class used by the
-- @OverloadedRecordFields@ extension.  See the
-- <https://gitlab.haskell.org/ghc/ghc/wikis/records/overloaded-record-fields
-- wiki page> for more details.
--
-----------------------------------------------------------------------------

module GHC.Records
       ( HasField(..)
       , getField
       , setField
       ) where

-- | Constraint representing the fact that the field @x@ can be get and set on
--   the record type @r@ and has field type @a@.  This constraint will be solved
--   automatically, but manual instances may be provided as well.
--
--   The function should satisfy the invariant:
--
-- > uncurry ($) (hasField @x r) == r

--   HasField :: forall {k}. k -> * -> * -> Constraint
--   getField :: forall {k} (x::k) r a. HasField x r a => r -> a
-- NB: The {k} means that k is an 'inferred' type variable, and
--     hence not provided in visible type applications.  Thus you
--     say     getField @"foo"
--     not     getField @Symbol @"foo"
class HasField x r a | x r -> a where
  -- | Function to get and set a field in a record.
  hasField :: r -> (a -> r, a)

-- | Selector function to extract the field from the record.
getField :: forall x r a . HasField x r a => r -> a
getField = snd . hasField @x

-- | Update function to set a new value of the field on the record.
setField :: forall x r a . HasField x r a => r -> a -> r
setField = fst . hasField @x
