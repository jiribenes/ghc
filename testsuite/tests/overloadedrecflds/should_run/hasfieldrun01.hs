{-# LANGUAGE DataKinds
           , DatatypeContexts
           , FlexibleInstances
           , GADTs
           , MultiParamTypeClasses
           , TypeFamilies
           , TypeApplications
  #-}
{-# OPTIONS_GHC -dcore-lint #-}

import GHC.Records (HasField(..), getField, setField)

type family B where B = Bool

data T = MkT { foo :: Int, bar :: B }

data U a b = MkU { baf :: a }
  deriving Show

data family V a b c d
data instance V x Int y [z] = MkVInt { baz :: (x, y, z, Bool) }

data W a where
  MkW :: { woo :: a } -> W [a]

data Eq a => X a = MkX { xoo :: a }
  deriving Show
data Y a = Eq a => MkY { yoo :: a }

data Z = MkZ1 { partial :: Int, total :: Bool }
       | MkZ2 { total :: Bool }
  deriving Show

t = MkT 42 True

u :: U Char Char
u = MkU 'x'

v = MkVInt (42, 'x', True, False)

w = MkW True

x = MkX True

y = MkY True

z = MkZ2 False

-- A virtual foo field for U
instance HasField "foo" (U a b) [Char] where
  hasField r = (const r, "virtual")

main = do print (getField @"foo" t)
          print (getField @"foo" (setField @"foo" t 11))
          print (getField @"bar" t)
          print (getField @"bar" (setField @"bar" t False))
          print (getField @"baf" u)
          print (setField @"baf" u 'y')
          print (getField @"foo" u)
          print (setField @"foo" u "ignored")
          print (getField @"baz" v)
          print (getField @"baz" (setField @"baz" v (40 :: Int, 'y', False, True)))
          print (getField @"woo" w)
          print (getField @"woo" (setField @"woo" w False))
          print (getField @"xoo" x)
          print (setField @"xoo" x False)
          print (getField @"yoo" y)
          print (getField @"yoo" (setField @"yoo" y False))
          print (getField @"total" (setField @"total" z True))
          print (setField @"partial" z 42)  -- Should throw a "No match" error
