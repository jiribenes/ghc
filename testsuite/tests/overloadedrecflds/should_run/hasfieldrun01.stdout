42
11
True
False
'x'
MkU {baf = 'y'}
"virtual"
MkU {baf = 'x'}
(42,'x',True,False)
(40,'y',False,True)
True
False
True
MkX {xoo = False}
True
False
True
